akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala

```diff
@@ -0,0 +1,209 @@
+        offset,
+        env.getPersistenceId,
+        env.getSequenceNr,
+        eventOption,
+        env.getTimestamp,
+        metaOption,
+        env.getEntityType,
+        env.getSlice)
+
+    case _ =>
+      fromStorageRepresentation(new String(bytes, UTF_8), manifest)
+  }
+
+  /**
+   * Deserialize an offset from a stored string representation and manifest.
+   * The offset is converted from its string representation to its real type.
+   */
+  private def fromStorageRepresentation(offsetStr: String, manifest: String): Offset = {
+    manifest match {
+      case TimestampOffsetManifest     => timestampOffsetFromStorageRepresentation(offsetStr)
+      case SequenceOffsetManifest      => Offset.sequence(offsetStr.toLong)
+      case TimeBasedUUIDOffsetManifest => Offset.timeBasedUUID(UUID.fromString(offsetStr))
+      case NoOffsetManifest            => NoOffset
+      case _ =>
+        val parts = manifest.split(':')
+        if (parts.length != 2)
+          throw new NotSerializableException(
+            s"Unimplemented deserialization of offset with manifest [$manifest] " +
+            s"in [${getClass.getName}]. [$manifest] doesn't contain two parts.")
+        val serializerId = parts(0).toInt
+        val serializerManifest = parts(1)
+        val bytes = Base64.getDecoder.decode(offsetStr)
+        serialization.deserialize(bytes, serializerId, serializerManifest).get match {
+          case offset: Offset => offset
+          case other =>
+            throw new NotSerializableException(
+              s"Unimplemented deserialization of offset with serializerId [$serializerId] and manifest [$manifest] " +
+              s"in [${getClass.getName}]. [${other.getClass.getName}] is not an Offset.")
+        }
+
+    }
+  }
+
+  /**
+   * Convert the offset to a tuple (String, String) where the first element is
+   * the String representation of the offset and the second element is its manifest.
+   */
+  private def toStorageRepresentation(offset: Offset): (String, String) = {
+    offset match {
+      case t: TimestampOffset => (timestampOffsetToStorageRepresentation(t), TimestampOffsetManifest)
+      case seq: Sequence      => (seq.value.toString, SequenceOffsetManifest)
+      case tbu: TimeBasedUUID => (tbu.value.toString, TimeBasedUUIDOffsetManifest)
+      case NoOffset           => ("", NoOffsetManifest)
+      case _ =>
+        val obj = offset.asInstanceOf[AnyRef]
+        val serializer = serialization.findSerializerFor(obj)
+        val serializerId = serializer.identifier
+        val serializerManifest = Serializers.manifestFor(serializer, obj)
+        val bytes = serializer.toBinary(obj)
+        val offsetStr = Base64.getEncoder.encodeToString(bytes)
+        (offsetStr, s"$serializerId:$serializerManifest")
+    }
+  }
+
+  private def timestampOffsetFromStorageRepresentation(str: String): TimestampOffset = {
+    try {
+      val parts = str.split(timestampOffsetSeparator)
+      if (parts.length == 4) {
```

Comment:
Not important, and subjective, but could perhaps be cleaner as a patmatch
```scala
parts match {
  case Array(ts, rts, seenPid, seenSeqnr) =>...
  case Array(ts) => 
  case Array(ts, rts) => 
  case lotsaparts =>
```

Reply:
I find this important. 😄